src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::fs::DirBuilder;
+use std::io::{Error, ErrorKind, Result};
+use std::os::unix::fs::DirBuilderExt;
+use std::path::{Path, PathBuf};
+
+use crate::{safe_join, SafePathBuf};
+
+/// Safe version of `DirBuilder` to protect from TOCTOU style of attacks.
+#[derive(Debug)]
+pub struct SafeDirBuilder {
+    root: PathBuf,
+    mode: u32,
+    recursive: bool,
+}
+
+impl SafeDirBuilder {
+    /// Creates a new set of options with default mode/security settings for all platforms and
+    /// also non-recursive.
+    pub fn new() -> Self {
+        SafeDirBuilder {
+            root: PathBuf::from("/"),
+            mode: 0o777,
+            recursive: false,
+        }
+    }
+
+    /// Indicates that directories should be created recursively, creating all parent directories.
+    /// Parents that do not exist are created with the same security and permissions settings.
+    pub fn recursive(&mut self, recursive: bool) -> &mut Self {
+        self.recursive = recursive;
+        self
+    }
+
+    /// Sets the mode to create new directories with. This option defaults to 0o777.
+    pub fn mode(&mut self, mode: u32) -> &mut Self {
+        self.mode = mode;
+        self
+    }
+
+    /// Sets the root directory for security check.
+    ///
+    /// # Safety
+    /// The `root` is the trusted part, and should not be modifiable by untrusted agents.
+    pub fn root<P: AsRef<Path>>(&mut self, root: P) -> &mut Self {
+        self.root = root.as_ref().to_path_buf();
+        self
+    }
+
+    /// Creates the specified directory with the options configured in this builder.
+    ///
+    /// The `path` must be a subdirectory of `SafePathBuf::root()`, otherwise error will be returned.
+    /// It is considered an error if the directory already exists unless recursive mode is enabled.
+    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<SafePathBuf> {
+        let mut root = self.root.canonicalize()?;
```

Comment:
IMO, validation of root should be done in `root`, but that will need to change its function signature to a Result type.

Or just for using a call-chain, from a user experience point, how about this style:

```rust
let sdb = SafeDirBuildere::new().root("/var/lib/sandbox-1").mode(0o777).build();
let sdb = sdb.unwarp();
```

Reply:
Changed the signature of `root()` to return error code.